### PR TITLE
Kezhan/change edge api to use index

### PR DESCRIPTION
**Pull request by linkerzhang**

### CODE REVIEW

File: include/onnxruntime/core/graph/graph.h [hunk 1]
```diff
@@ -66,13 +67,18 @@ class Node {
     /** Gets the Node that this EdgeEnd refers to. */
     const Node& GetNode() const noexcept;
 
-    /** Gets the NodeArg that this edge end refers to.
-    @returns NodeArg pointer or nullptr if this is a control edge. */
-    const NodeArg* GetNodeArg() const noexcept;
+    /** Gets the source slot index.
+	@returns the source slot index of <*this> edge.*/
+    int GetSourceSlotIndex() const;
```
Comment (weixingzhang) [inline include/onnxruntime/core/graph/graph.h]: Is it better to change it to GetSrcSlotIndex to align with GetDstSlotIndex below?

File: include/onnxruntime/core/graph/graph.h [hunk 2]
```diff
@@ -54,9 +54,10 @@ class Node {
     Construct an EdgeEnd
     @param node The source node if this is an input edge to the current node, 
     or the destination node if this is an output edge from the current node.
-    @param node_arg The NodeArg to use for the edge.
+    @param src_slot_index The node arg slot of source node of the edge.
+	@param dst_slot_index The node arg slot of destination node of the edge.
     */
-    EdgeEnd(const Node& node, const NodeArg& node_arg) noexcept;
+    EdgeEnd(const Node& node, int src_slot_index, int dst_slot_index) noexcept;
```
[the remark this answers is not in the document]
Comment (yuanbyu) [reply include/onnxruntime/core/graph/graph.h]: No. It is still a node end unless we change it to <srcNode, srcIndex, dstNode, dstIndex>.

File: include/onnxruntime/core/graph/graph.h (hunk 1 above)
Comment (yuanbyu) [reply include/onnxruntime/core/graph/graph.h]: We probably should just call it GetSrcNodeArgIndex and GetDstNodeArgIndex?

File: include/onnxruntime/core/graph/graph.h [hunk 3]
```diff
@@ -779,7 +792,7 @@ class Graph {
   struct ResolveContext {
     ResolveContext() = default;
 
-    std::unordered_map<std::string, Node*> output_args;
+    std::unordered_map<std::string, std::pair<Node*, int>> output_args;
```
[the remark this answers is not in the document]
Comment (yuanbyu) [reply include/onnxruntime/core/graph/graph.h]: If I understand it correctly, I am actually fine with the original code. A comment would be helpful.

File: onnxruntime/core/graph/graph.cc [hunk 4]
```diff
@@ -783,74 +791,84 @@ const NodeArg* Graph::GetNodeArgIncludingParentGraphs(const std::string& node_ar
...
     }
   }
-  ONNXRUNTIME_ENFORCE(valid);
-  nodes_[dst_node_index]->MutableRelationships().input_edges.insert(Node::EdgeEnd(*nodes_[src_node_index], node_arg));
-  nodes_[src_node_index]->MutableRelationships().output_edges.insert(Node::EdgeEnd(*nodes_[dst_node_index], node_arg));
+  if (nullptr == dst_arg) {
+    ONNXRUNTIME_THROW("Invalid destination node arg slot specified when adding edge.");
+  }
+
+  if (src_arg != dst_arg && src_arg->Type() != dst_arg->Type()) {
```
[the remark this answers is not in the document]
Comment (kkaranasos) [reply onnxruntime/core/graph/graph.cc]: You are right, the check is correct. 
But still, shouldn't we update the dst_arg to be the same as the src_arg when the edge is added?

File: include/onnxruntime/core/graph/graph.h (hunk 1 above)
Comment (linkerzhang) [reply include/onnxruntime/core/graph/graph.h]: Sure. Thanks!

**Pull request merged**